sycl/include/sycl/vector_preview.hpp

```diff
@@ -178,222 +104,102 @@ using rel_t = typename std::conditional_t<
-// Single element byte. Multiple elements will propagate through a later
-// specialization.
-template <> struct VecStorage<std::byte, 1, void> {
-  using DataType = std::int8_t;
-#ifdef __SYCL_DEVICE_ONLY__
-  using VectorDataType = std::int8_t;
-#endif // __SYCL_DEVICE_ONLY__
-};
-#endif // (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
+      std::is_same_v<T, std::byte>, std::uint8_t,
+#else
+      false, T,
+#endif
+      typename std::conditional_t<std::is_same_v<T, bool>, std::int8_t,
+      typename std::conditional_t<std::is_same_v<T, sycl::half>, sycl::detail::half_impl::StorageT,
+      typename std::conditional_t<std::is_same_v<T, sycl::ext::oneapi::bfloat16>,
+      sycl::ext::oneapi::detail::Bfloat16StorageT, T>>>>;
 
-// Single element signed integers
-template <typename T>
-struct VecStorage<T, 1, typename std::enable_if_t<is_sigeninteger_v<T>>> {
-  using DataType = T;
-#ifdef __SYCL_DEVICE_ONLY__
-  using VectorDataType = DataType;
-#endif // __SYCL_DEVICE_ONLY__
-};
+// Helper functions to convert to and from OpenCL type.
+template <typename T, typename U> constexpr auto VecConvertFromOpenCLType(U value) {
 
-// Single element unsigned integers
-template <typename T>
-struct VecStorage<T, 1, typename std::enable_if_t<is_sugeninteger_v<T>>> {
-  using DataType = T;
 #ifdef __SYCL_DEVICE_ONLY__
-  using VectorDataType = DataType;
-#endif // __SYCL_DEVICE_ONLY__
-};
+  // Convert from std::array to ext_vector_type.
+  // Check if T is std::array and U is ext_vector_type. If so, you sycl::bit_cast<T>.
+  if constexpr (detail::is_std_array_v<T> && detail::is_ext_vector_type_v<U>)
+    return sycl::bit_cast<T>(value);
+  else
+#endif
+    if constexpr (std::is_same_v<T, sycl::ext::oneapi::bfloat16>)
+      return sycl::bit_cast<T>(value);
+    else if constexpr (std::is_same_v<T, bool>)
+      return static_cast<int8_t>(value);
+    else
+      return static_cast<T>(value);
+}
 
-// Single element floating-point (except half/bfloat16)
-template <typename T>
-struct VecStorage<
-    T, 1,
-    typename std::enable_if_t<!is_half_or_bf16_v<T> && is_sgenfloat_v<T>>> {
-  using DataType = T;
-#ifdef __SYCL_DEVICE_ONLY__
-  using VectorDataType = DataType;
-#endif // __SYCL_DEVICE_ONLY__
-};
-// Multiple elements signed/unsigned integers and floating-point (except
-// half/bfloat16)
-template <typename T, int N>
-struct VecStorage<
-    T, N,
-    typename std::enable_if_t<isValidVectorSize(N) &&
-                              (is_sgeninteger_v<T> ||
-                               (is_sgenfloat_v<T> && !is_half_or_bf16_v<T>))>> {
-  using DataType =
-      typename VecStorageImpl<typename VecStorage<T, 1>::DataType, N>::DataType;
-#ifdef __SYCL_DEVICE_ONLY__
-  using VectorDataType =
-      typename VecStorageImpl<typename VecStorage<T, 1>::DataType,
-                              N>::VectorDataType;
-#endif // __SYCL_DEVICE_ONLY__
-};
+template <typename T, typename U = void> constexpr auto VecConvertToOpenCLType(T value) {
```

Comment:
Is SFINAE intentional as a customization point for some other types? If so needs a comment, if not, just drop the `U`. Also, I'd prefer `To`/`From` as type parameters' names.

And also guard with `__SYCL_DEVICE_ONLY__`.

Also, what is a "legal" `From` type here? Is that somehow limited to types supported by `sycl::vec`? If so, can we check for that?